### PR TITLE
Add ability to register as guest

### PR DESCRIPTION
**Pull request by neosam**
This pull request will add a method called guest_session.  It will automatically set the session attribute in the Client struct. Additionally there is an example which runs the method to test if it is actually working.

### CODE REVIEW

**Comment by jimmycuadra** [comment]
Thanks for the contribution! We'll get this reviewed as soon as we can.

File: src/lib.rs [hunk 1]
```diff
@@ -73,4 +76,24 @@ impl Client {
...
+            r0::account::register::Endpoint::request_path(()));
+        let mut response = self.hyper
+            .request(r0::account::register::Endpoint::method().into_hyper(),
+                self.homeserver_url.join(&full_url)?)
+            .body("{}")
+            .send()?;
+        let mut response_str = String::new();
+        response.read_to_string(&mut response_str)?;
+        let user_session: r0::account::register::Response = serde_json::from_str(&response_str)?;
+        let user_id = UserId::try_from(&user_session.user_id.clone())?;
```
Comment (vberger) [inline src/lib.rs]: This is a bug of ruma-client-api, here `register::Response.user_id` should already be of type `UserId` (my bad, I missed it in my PR about spec inconsistencies)

Comment (vberger) [inline src/lib.rs]: Also, as you completely move things out of `user_session`, the `.clone()` calls should not be needed

Comment (neosam) [inline src/lib.rs]: Thanks for the feedback.  You're right, I removed the .clone()s.

**Comment by jimmycuadra** [comment]
I'm not sure how I feel about adding a method to the client that provides a specific user flow like "register as a guest" that is not exposed as a unique endpoint in the Matrix API. I had imagined ruma-client to be more of a simple combination of ruma-client-api and hyper, with the necessary data structures like `Session` to make it minimally ergonomic. I'm concerned that if we add APIs like this to ruma-client, we'll end up with too many of these high level methods and it'll be hard for users to figure out which methods to use for a particular task, or methods that are too narrow of a use case.

@ruma/contributors, thoughts?

**Comment by neosam** [comment]
What do you think about transforming it into a general register method?

**Comment by jimmycuadra** [comment]
I think that'd be a better way to go. I'm also not sure how much we should expose the types from ruma-client-api directly. There are a few ways it might work:

* Expose the ruma-client-api types directly. The user then creates values for any required `BodyParams` or `QueryParams` and passes them to the ruma-client method.
* Use positional arguments on the ruma-client method and construct `BodyParams` and `QueryParams` as part of the method's implementation.
* Create a new public "options" struct for each method in ruma-client to avoid having a long list of positional arguments, and once again converting these options structs into the relevant ruma-client-api types inside the method body. This could help paper over some of the awkward Matrix APIs, like the fact that in the account registration endpoint, "kind" is a query string parameter, but all the other parameters are part of the body.

Sorry to have to discuss this on your PR, but we only just started spiking out ruma-client so we hadn't yet come to decisions on the basic design of the library. You're a very early adopter. :}

**Comment by vberger** [comment]
I'm not completely sure about what to do. In my opinion it's good to not expose the ruma-client-api types directly, splits BodyParams/QueryParams/PathParams are indeed awkward.

However, using only positional arguments or options struct seem extreme, when there are respectively many or few arguments to the request.

Maybe a course of action could be to have positional parameters for the fundamental arguments of a query, and stuff all optional/secondary values in an Options struct (it can be reasonable to use with the `{ ..Default::default() }` syntax.

Doing things intelligently would allow use to improve substantially on the HTTP API. First example coming to mind is the API linked to state values. All endpoints are duplicated to separate between the empty state key and regular state keys. We could cleanly merge them back using an Option-like enum:

```rust
pub enum StateKey {
    Key(String),
    EmptyKey
}
```

**Comment by jimmycuadra** [comment]
@vberger, that sounds like a good approach. The other rule of thumb we might want to follow is to also use a struct when there are more than a few required parameters, because function signatures tend to get confusing when you have 3+ positional arguments. It's harder, or simply less obvious, which order they go in, so a struct is more useful in cases like that, because the fields are named.

I do worry about APIs that have a lot of different data structures like /search. If someone is doing a complicated search that uses all the bells and whistles, there will be several different types they'll have to weave together to pass to the ruma-client method, and it would feel very heavy handed to create special versions of all of them in ruma-client just to avoid exposing ruma-client-api types.

**Comment by vberger** [comment]
> I do worry about APIs that have a lot of different data structures like /search.

I would say, the `***Params` and `Response` types should not be exposed, but additional structures they use could be. In this cass they should properly be reexported in ruma-client.

**Comment by jimmycuadra** [comment]
You don't think the response types should be reexported? I don't think there'd be any difference between the ruma-client-api type and anything else we could come up with. I think you're right on the various request types.

**Comment by vberger** [comment]
Hmm indeed, I've written too quickly without thinking. There is probably no real reason to not expose Response, expect the confusing fact that all methods will return a struct named `Response`, while this will always be a different type.

This can be clarified with renaming-reexports though.

**Comment by vberger** [comment]
And it will require a good documentation effort on ruma-client-api, if the goal is to have ruma-client being usable stand-alone, without having to worry about its dependencies.

Mainly, change the docs `Response` for each endpoint to something more helpfull than "This endpoint's response", as they would appear in ruma-client's docs, out of the context of their endpoint module.

**Comment by vberger** [comment]
I wanted to try and see how this could be applied to the register endpoint, to give a concrete example, but... It seems there are a few things left to be fixed for this endpoint... Typically, guest accounts don't require a password, do they? (the spec seems to say that they do, which I find strange)

**Comment by jimmycuadra** [comment]
You're right, that is strange. @ara4n or @richvdh, can you weigh in on the password requirement for guest accounts?

**Comment by vberger** [comment]
Looks like there is already a bug opened about this question: https://github.com/matrix-org/matrix-doc/issues/771

**Comment by jplatte** [comment]
Yeah, the register endpoint is a complete mess, despite being quite simple. There's another bug report about kind being a query parameter instead of a body parameter, which I think was discussed on some issue here or in the ruma IRC / Matrix channel too: https://github.com/matrix-org/matrix-doc/issues/672

I like the idea to mainly glue ruma-client-api and hyper together in a simple manner, while also hiding some of the dirty details such as the distinction between query, body and path parameters and the fact that you might have to violate the spec for some operations :D (although I do hope that that's gonna be fixed with the next, probably-to-be-delayed-because-of-current-federation-meltdowns version)

**Comment by jimmycuadra** [comment]
Ha, yeah, registration/auth is pretty awkward in Matrix. matrix-org/matrix-doc#672 was actually reported by me, dunno why it says Rich did. Registration was the first thing I worked on when I started Ruma and it took a lot of wrangling the spec to figure it out. Some of the things I reported got fixed/improved, but it's still got problems.

**Comment by richvdh** [comment]
> matrix-org/matrix-doc#672 was actually reported by me, dunno why it says Rich did.

I think you are thinking of https://matrix.org/jira/browse/SPEC-385.

**Comment by neosam** [comment]
I try to summarize the discussion until now.  This is what I understood in which direction we approximately want to go:

- Make the code as simple as possible, only connect ruma-client-api and hyper
- Use structs from ruma-client-api except for the ***Params.
- If more than three attributes would be required, use a custom struct as parameter to avoid long method signatures.

In this case we would just have a register method which requires these parameters:

- kind
- username
- password
- bind_email
- auth

As a user of this method I would find it useful if I could pass something like this because I don't need anything when I register as a guest:
```rust
enum RegisterParam {
  Guest,
  User {
    username: String,
    password: String,
    bind_email: bool,
    auth: WhatEver
  }
}
```

Then I could write:
```rust
let mut client = Client::new("https://matrix.org/")
       .expect("Could not connect to server");
client.register(RegisterParam::Guest)
       .expect("Cound not register as guest");
```
or
```rust
let mut client = Client::new("https://matrix.org/")
       .expect("Could not connect to server");
let register_param = RegisterParam::User {
  username: "rumeuser".to_string(),
  password: "secret".to_string(),
  bind_email: false,
  Auth: WhatEver::new()
}
client.register(register_param)
       .expect("Cound not register as guest");
```

I hope there are not too many syntax errors in the dummy code :)

**Comment by neosam** [comment]
Should register(...) then replace the session attribute in Client or just return the Respone?  Or both?

**Comment by vberger** [comment]
Also, cc https://github.com/ruma/ruma-client-api/pull/13, we might want to wait on this?

**Comment by neosam** [comment]
@jimmycuadra: Do I understand it right that we do not need a persistent connection to the server because the protocol runs over http(s)?  Then I would say we could just return something like a `MatrixSession` because we can login or register multiple times with the `Client`.  I guess this is more flexible and fits better to the API.

I belief everything has the potential to get confusing :)

**Comment by jimmycuadra** [comment]
HTTP keep-alive will improve performance, but no, there isn't any sort of "always on" connection needed to the server. Everything is the usual stateless HTTP requests. At least this is how it exists in Synapse and Ruma today. It's possible if not likely that future transports like WebSockets will be added.

**Comment by vberger** [comment]
Just noting that https://github.com/ruma/ruma-client-api/pull/13 is merged. You can update on it @neosam, so you can directly build a sane interface that should _mostly_ be compatible with what synapse does currently.

**Comment by jimmycuadra** [comment]
We might actually not even need a `Client` type, if we end up using `Session` as the primary interface. We could just have a few top-level functions that are used to create `Session`s, e.g. `ruma_client::{login, logout, register}`. The other question is how to handle other API endpoints that don't require authentication. There should be a way to use those without a `Session`, but we don't want a proliferation of top-level functions. On the other hand, if you *do* have a `Session`, it would be awkward to use one type for endpoints requiring auth and one for endpoints not requiring auth. If there is some other "unauthenticated" type that contains all methods for all those API calls, `Session` could include the same methods and just call the other type's methods internally. Or maybe we should actually enumerate all the endpoints that don't require auth to see if top-level functions for all of them would be feasible.

**Comment by jimmycuadra** [comment]
That actually seems pretty workable to make them all top-level functions, especially considering those endpoints that probably do require auth and are just not labeled correctly in the spec.

**Comment by jimmycuadra** [comment]
I've done some reworking of the code to use the Tokio (async) branch of Hyper, so whatever we decide here, this'll need some reworking.

@jplatte also brought up a good point, that if all the unauthenticated API calls are done from top-level methods, then it would be less ergonomic to share the same Hyper client for all calls. For that reason, I think we should go back to keeping session state within `Client` and using `Client` for all API calls.

**Comment by neosam** [comment]
@jimmycuadra Thanks for the update.  I will now pull the changes and try to fix the pull request.

**Comment by neosam** [comment]
I think it makes sense to close this pull request and to start a new one.  What do you think?

**Comment by vberger** [comment]
@neosam I think it's ok to do that, if it's simpler for you.

**Comment by vberger** [comment]
Maybe, just don't close this PR for now (even if you create an other one), so we can keep this discussion around easily.

**Comment by jimmycuadra** [comment]
Closing this, as the questions discussed here have since been answered.

**Pull request closed without merging**